src/compiler/program.ts

```diff
@@ -107,54 +107,80 @@ namespace ts {
 
-        function writeFileIfUpdated(fileName: string, data: string, writeByteOrderMark: boolean): void {
+        function hasFingerprint(fileName: string, hash: string, byteOrderMark: boolean, mtime: Date) {
+            const fingerprint = outputFingerprints && outputFingerprints.get(fileName);
+            // If output has not been changed, and the file has no external modification
+            return fingerprint &&
+                fingerprint.byteOrderMark === byteOrderMark &&
+                fingerprint.hash === hash &&
+                fingerprint.mtime.getTime() === mtime.getTime();
+        }
+
+        function setFingerprint(fileName: string, hash: string, byteOrderMark: boolean, mtime: Date) {
             if (!outputFingerprints) {
                 outputFingerprints = createMap<OutputFingerprint>();
             }
 
-            const hash = system.createHash!(data); // TODO: GH#18217
-            const mtimeBefore = system.getModifiedTime!(fileName); // TODO: GH#18217
+            outputFingerprints.set(fileName, {
+                hash,
+                byteOrderMark,
+                mtime,
+            });
+        }
 
-            if (mtimeBefore) {
-                const fingerprint = outputFingerprints.get(fileName);
-                // If output has not been changed, and the file has no external modification
-                if (fingerprint &&
-                    fingerprint.byteOrderMark === writeByteOrderMark &&
-                    fingerprint.hash === hash &&
-                    fingerprint.mtime.getTime() === mtimeBefore.getTime()) {
-                    return;
-                }
+        function writeFileIfUpdated(writeFileCallback: WriteFileCallback, fileName: string, data: string, writeByteOrderMark: boolean) {
+            if (!system.createHash || !system.getModifiedTime) {
+                writeFileCallback(fileName, data, writeByteOrderMark);
+                return;
             }
 
-            system.writeFile(fileName, data, writeByteOrderMark);
+            const hash = system.createHash(data);
 
-            const mtimeAfter = system.getModifiedTime!(fileName) || missingFileModifiedTime; // TODO: GH#18217
+            const mtimeBefore = system.getModifiedTime(fileName);
+            if (mtimeBefore && hasFingerprint(fileName, hash, writeByteOrderMark, mtimeBefore)) {
+                return;
+            }
 
-            outputFingerprints.set(fileName, {
-                hash,
-                byteOrderMark: writeByteOrderMark,
-                mtime: mtimeAfter
-            });
+            writeFileCallback(fileName, data, writeByteOrderMark);
+
+            const mtimeAfter = system.getModifiedTime(fileName) || missingFileModifiedTime;
+
+            setFingerprint(fileName, hash, writeByteOrderMark, mtimeAfter);
+        }
+
+        async function writeFileIfUpdatedAsync(writeFileAsyncCallback: WriteFileAsyncCallback, fileName: string, data: string, writeByteOrderMark: boolean) {

```

Comment:
I couldn't figure out how to share more of this code with the sync version.  Is there a standard pattern?